### ISSUE TITLE
一块V100训练spider数据集只需2个多小时？

### ISSUE THREAD

**Issue by whdata08** [DreamerDeo/UniSAr_text2sql#2]
用的默认的参数，total_num_update=10000，max_tokens 1024 。一块V100两个多小时就训练好了，请问一下这是否正常。因为看您论文中备注的是4块V100训练10小时。

**Comment by DreamerDeo** [comment]
Hi, 感谢对我们工作的关注！
实验部分，我用的是v100 16g, 训练数据量确定是8700条吗？能不能发一下log～

**Comment by whdata08** [comment]
您好，十分感谢您的回复！
log请见附件，训练数据是只用了train_spider.json的7000条数据。(total_num_update=10000，max_tokens=1024，均用的github仓库代码中的默认参数）&nbsp;


训练tensorboard如下。

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;









------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "DreamerDeo/UniSAr_text2sql"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月8日(星期五) 上午10:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [DreamerDeo/UniSAr_text2sql] 一块V100训练spider数据集只需2个多小时？ (Issue #2)





 
Hi, 感谢对我们工作的关注！
 实验部分，我用的是v100 16g, 训练数据量确定是8700条吗？能不能发一下log～

**Comment by whdata08** [comment]
[log.txt](https://github.com/DreamerDeo/UniSAr_text2sql/files/8449158/log.txt)
上条回复是回复的邮件，这里把log.txt重新发送以下。用的1块V100，32G。

**Comment by DreamerDeo** [comment]
[unisar_log_4xV100-32G.txt](https://github.com/DreamerDeo/UniSAr_text2sql/files/8449686/unisar_log_4xV100-32G.txt)

我去查了一下，根据 [这个issue](https://github.com/pytorch/fairseq/issues/2105) `total_num_update=10000` 是指每块GPU上都训练10000步（总共40000步），整体参数的更新步数应该看epoch数。
PS: fairseq中 batch size 和 max_tokens 这两个参数都可以决定真正的batch-size，但是一般倾向于用后者，因为这样fairseq可以更好地动态计算batch-size。

上面的文件是我在实验早期用4块32g V100训练的log，可以看到训10000步花了差不多4个小时（这个是因为我在集群上跑的8700数据，可能因为多卡通信或者服务器配置之类的问题，会比你的两个小时长一些）。
整体更新步数，我的epoch是跑到了`"epoch": 76`, 这个差不多是你的 `"epoch": 14` 的 3～4倍 （根据log，我这里的bart是从epoch 42开始的，而你的bart从epoch 1开始的了）。

**Comment by whdata08** [comment]
您好：
     我认真的看了您的回复，并花了时间尝试自己去找原因，但是有几个疑惑一直想不清楚。
现状：模型训练1小时左右验证集loss最优，后续会上升。见文末tensorboard图。
问题：
1. 您的bart是从epoch 42开始训练的，而我的在加载模型时也会提示'Loaded checkpoint ./data/BART-large/model.pt (epoch 41 @ 0 updates)'，bart.large模型下载地址为
[https://github.com/pytorch/fairseq/blob/main/examples/bart/README.md)中的bart.large](url)
     a.首先不清楚为什么我下载的预训练模型会有 (epoch 41 @ 0 updates)的提示
     b.其次您的模型从epoch42开始，我的从1开始，这里是不是有问题呢?还是说您已经训练了41轮，日志中是接着上次的41轮继 
     续训练的
2.您的训练日志中："total": "8358.4","bsz": "29.5",这是否意味着每条数据长度为total/bsz=283
    而我的训练日志中："total": "378.8","bsz": "10",每条数据长度约为38。
    (我理解的目标sql语句每条38的长度应该是正常的吧，不清楚为什么我们的差距这么大。)
描述的问题有点多，如果您有时间回复我其中部分问题，我将不胜感激。
![loss](https://user-images.githubusercontent.com/87259800/162617723-60454f2e-1685-434f-b0d0-d4a4845d7ee9.png)
![acc](https://user-images.githubusercontent.com/87259800/162617747-883ab188-67d3-440d-811e-7abedf5257f0.png)

**Comment by DreamerDeo** [comment]
_**预训练模型会有 (epoch 41 @ 0 updates)的提示**_
BART这个checkpoint应该在pretrain阶段训练了41个epoch。我猜从optimizer或者dataloader里能读出这个信息来

_**模型从epoch42开始**_
这个取决于加不加`--reset-dataloader`参数，我上面贴的这个log是早起实验没加的，后来的实验包括release的这个checkpoint都是加的。实验结果表明除了start epoch这个读数以外并没有发现什么明显的区别
<img width="674" alt="image" src="https://user-images.githubusercontent.com/9781506/162665070-a244ab55-9b96-49e6-be24-f4d2bb98300a.png">

[来源：fairseq手册](https://fairseq.readthedocs.io/en/latest/command_line_tools.html#fairseq-train)


_**"wpb": "8358.4","bsz": "29.5"**_
这份log对应的早期实验，输入输出和最终的版本不太一致，输出还包括一些附加的内容，所以会长很多～我重新在服务器上跑了一下，和你的数据是类似的

PS: loss和bleu跟sql exact-match有相关性（1、2变好，3有可能变好）但是不绝对成正比，即best checkpoint是根据bleu选出来的，而bleu则是translation这个task的metric。所以为了挑出最好的text2sql ckpt，需要（1）用fairseq重新定义一个parsing task 或者（2）遍历evaluate所有的ckpt

**Comment by DreamerDeo** [comment]
closing due to inactivity. please reopen if this problem reappears.
